### PR TITLE
Disallow several `indexed` attributes for the same event parameter

### LINKED ISSUE #13681: The `indexed` modifier can be specified multiple times for event arguments
## Description

```solidity
pragma solidity ^0.8.17;

interface Foo {}

interface Bar {
    event MyEvent(Foo indexed indexed indexed some);
}
```

The above contract is _compiled_ while my expectation was to get a parse error. Other tools are failing to parse such source. For an instance, `solidity-coverage` HardHat plugin is failing on it. Is it intended to use multiple `indexed` entries for event arguments? Also is it documented somewhere?

## Environment

This is reproducible in Remix IDE.

## Steps to Reproduce

Run code sample from "Description" section in Remix IDE.

Regards.

### PR DESCRIPTION
**Pull request by vlad-khramov**
Fixes #13681

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ Bugfixes:
  * SMTChecker: Fix internal error on chain assignments using static fully specified state variables.
  * SMTChecker: Fix internal error when using user defined types as mapping indices or struct members.
  * SMTChecker: Fix internal error when deleting struct member of function type.
+ * Parser: Disallow several ``indexed`` attributes for the same event parameter.
```
Comment (nikola-matic) [inline Changelog.md]:
```suggestion
 * Parser: Disallow multiple ``indexed`` attributes for the same event parameter.
```

**Review by ekpyron — COMMENTED** [review]
LGTM apart from https://github.com/ethereum/solidity/pull/13816#discussion_r1053345712

**Comment by cameel** [comment]
Just one thing. We should try to "check around the edges" and increase test coverage when fixing cases like this. I.e. here, if I were fixing it myself, I'd also make sure to check if any other modifiers/keywords can be used twice (i.e. `anonymous`, `global`, `virtual`, `override`, `constant`, `immutable`, `pure`, `view`, `payable`, `memory`, `calldata`, `storage`, `external`, `public`, `internal`, `private`, `experimental`) and if we have test cases covering these cases. `indexed` might not be the only thing that slipped through.

File: libsolidity/parsing/Parser.cpp [hunk 2]
```diff
@@ -785,7 +785,12 @@ ASTPointer<VariableDeclaration> Parser::parseVariableDeclaration(
 		else
 		{
 			if (_options.allowIndexed && token == Token::Indexed)
+			{
+				if (isIndexed)
+					parserError(5399_error, string("Indexed already specified."));
```
Comment (cameel) [inline libsolidity/parsing/Parser.cpp]:
```suggestion
					parserError(5399_error, "Indexed already specified.");
```

**Comment by vlad-khramov** [comment]
intermediate commit, work still in progress

**Comment by vlad-khramov** [comment]
- Added or checked tests for multiple usage for all keywords (see table below)
- Removed ability to add `indexed` for parameters of modifier 
- Added or checked tests for all invalid keywords usage for variable definitions in parameter list + in single variable definitions.


| keyword      | type     | where         | new | test                                                                    |
|--------------|----------|---------------|:---:|-------------------------------------------------------------------------|
| indexed      | invalid  | mod param     |  +  | modifiers/invalid_parameter_indexed.sol                                 |
|              | invalid  | fn par        |  +  | parsing/invalid_function_parameter_and_return_var_indexed.sol           |
|              | invalid  | fn ret        |  +  | parsing/invalid_function_parameter_and_return_var_indexed.sol           |
|              | multiple | ev param      |  +  | events/event_multiple_indexed_of_one_parameter.sol                      |
|              | invalid  | err param     |  +  | errors/invalid_parameter_indexed.sol                                    |
|              | invalid  | try ret       |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_indexed.sol           |
|              | invalid  | catch param   |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_indexed.sol           |
|              | invalid  | file var      |  +  | parsing/invalid_variable_indexed.sol                                    |
|              | invalid  | state var     |  +  | parsing/invalid_variable_indexed.sol                                    |
|              | invalid  | fn var        |  +  | parsing/invalid_variable_indexed.sol                                    |
| anonymous    | multiple | mod           |  +  | events/anonymous_event_multiple_anonymous_specifiers.sol                |
| global       | multiple | using         |  +  | using/global_multiple_global_keywords.sol                               |
| virtual      | multiple | fn, mod       |     | inheritance/virtual/duplicate.sol                                       |
| override     | multiple | mod           |     | parsing/multiple_modifier_overrides.sol                                 |
|              | multiple | state var, fn |     | inheritance/override/detect_double_override.sol                         |
| constant     | invalid  | mod param     |  +  | modifiers/invalid_parameter_mutability.sol                              |
|              | invalid  | fn par        |     | nameAndTypeResolution/105_constant_input_parameter.sol                  |
|              | invalid  | fn ret        |  +  | parsing/invalid_function_parameter_and_return_var_mutability.sol        |
|              | invalid  | ev param      |  +  | events/invalid_parameter_mutability.sol                                 |
|              | invalid  | err param     |  +  | errors/invalid_parameter_mutability.sol                                 |
|              | invalid  | try ret       |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_mutability.sol        |
|              | invalid  | catch param   |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_mutability.sol        |
|              | multiple | file var      |  +  | constants/file_level_multiple_constant_keywords.sol                     |
|              | multiple | state var     |  +  | parsing/multiple_mutability_specifiers.sol                              |
|              | invalid  | fn var        |  +  | parsing/invalid_variable_mutablity.sol                                  |
| immutable    | invalid  | mod param     |  +  | modifiers/invalid_parameter_mutability.sol                              |
|              | invalid  | fn par        |     | immutable/as_function_param.sol                                         |
|              | invalid  | fn ret        |  +  | parsing/invalid_function_parameter_and_return_var_mutability.sol        |
|              | invalid  | ev param      |  +  | events/invalid_parameter_mutability.sol                                 |
|              | invalid  | err param     |  +  | errors/invalid_parameter_mutability.sol                                 |
|              | invalid  | try ret       |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_mutability.sol        |
|              | invalid  | catch param   |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_mutability.sol        |
|              | invalid  | file var      |     | constants/immutable_at_file_level.sol                                   |
|              | multiple | state var     |  +  | parsing/multiple_mutability_specifiers.sol                              |
|              | invalid  | fn var        |  +  | parsing/invalid_variable_mutablity.sol                                  |
| public       | invalid  | mod param     |  +  | modifiers/invalid_parameter_visibility.sol                              |
|              | invalid  | fn par        |  +  | visibility/invalid_function_parameter_and_return_var_visibility.sol     |
|              | invalid  | fn ret        |  +  | visibility/invalid_function_parameter_and_return_var_visibility.sol     |
|              | invalid  | ev param      |  +  | events/invalid_parameter_visibility.sol                                 |
|              | invalid  | err param     |  +  | errors/invalid_parameter_visibility.sol                                 |
|              | invalid  | try ret       |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_visibility.sol        |
|              | invalid  | catch param   |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_visibility.sol        |
|              | invalid  | file var      |  +  | visibility/file_level_variable_visibility.sol                           |
|              | multiple | fn            |  +  | visibility/multiple_visibility_specifiers.sol                           |
|              | multiple | state var     |  +  | visibility/multiple_visibility_specifiers.sol                           |
|              | invalid  | fn var        |  +  | visibility/invalid_variable_visibility.sol                              |
| internal     | invalid  | mod param     |  +  | modifiers/invalid_parameter_visibility.sol                              |
|              | invalid  | fn par        |  +  | visibility/invalid_function_parameter_and_return_var_visibility.sol     |
|              | invalid  | fn ret        |  +  | visibility/invalid_function_parameter_and_return_var_visibility.sol     |
|              | invalid  | ev param      |  +  | events/invalid_parameter_visibility.sol                                 |
|              | invalid  | err param     |  +  | errors/invalid_parameter_visibility.sol                                 |
|              | invalid  | try ret       |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_visibility.sol        |
|              | invalid  | catch param   |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_visibility.sol        |
|              | invalid  | file var      |  +  | visibility/file_level_variable_visibility.sol                           |
|              | multiple | fn            |  +  | visibility/multiple_visibility_specifiers.sol                           |
|              | multiple | state var     |  +  | visibility/multiple_visibility_specifiers.sol                           |
|              | invalid  | fn var        |  +  | visibility/invalid_variable_visibility.sol                              |
| private      | invalid  | mod param     |  +  | modifiers/invalid_parameter_visibility.sol                              |
|              | invalid  | fn par        |  +  | visibility/invalid_function_parameter_and_return_var_visibility.sol     |
|              | invalid  | fn ret        |  +  | visibility/invalid_function_parameter_and_return_var_visibility.sol     |
|              | invalid  | ev param      |  +  | events/invalid_parameter_visibility.sol                                 |
|              | invalid  | err param     |  +  | errors/invalid_parameter_visibility.sol                                 |
|              | invalid  | try ret       |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_visibility.sol        |
|              | invalid  | catch param   |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_visibility.sol        |
|              | invalid  | file var      |  +  | visibility/file_level_variable_visibility.sol                           |
|              | multiple | fn            |  +  | visibility/multiple_visibility_specifiers.sol                           |
|              | multiple | state var     |  +  | visibility/multiple_visibility_specifiers.sol                           |
|              | invalid  | fn var        |  +  | visibility/invalid_variable_visibility.sol                              |
| external     | invalid  | mod param     |  +  | modifiers/invalid_parameter_visibility.sol                              |
|              | invalid  | fn par        |  +  | visibility/invalid_function_parameter_and_return_var_visibility.sol     |
|              | invalid  | fn ret        |  +  | visibility/invalid_function_parameter_and_return_var_visibility.sol     |
|              | invalid  | ev param      |  +  | events/invalid_parameter_visibility.sol                                 |
|              | invalid  | err param     |  +  | errors/invalid_parameter_visibility.sol                                 |
|              | invalid  | try ret       |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_visibility.sol        |
|              | invalid  | catch param   |  +  | tryCatch/invalid_returns_vars_and_catch_parameter_visibility.sol        |
|              | invalid  | file var      |  +  | visibility/file_level_variable_visibility.sol                           |
|              | multiple | fn            |  +  | visibility/multiple_visibility_specifiers.sol                           |
|              | multiple | state var     |  +  | visibility/multiple_visibility_specifier_external.sol                   |
|              | invalid  | fn var        |  +  | visibility/invalid_variable_visibility.sol                              |
| payable      | multiple | mod param     |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | mod param     |  +  | modifiers/elementary_non_address_state_mutability_modifier_argument.sol |
|              | multiple | fn par        |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | fn par        |     | parsing/elementary_non_address_state_mutability_argument.sol            |
|              | multiple | fn ret        |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | fn ret        |     | parsing/elementary_non_address_state_mutability_return.sol              |
|              | multiple | ev param      |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | ev param      |  +  | events/elementary_non_address_state_mutability_event_argument.sol       |
|              | multiple | err param     |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | err param     |  +  | errors/elementary_non_address_state_mutability_error_argument.sol       |
|              | multiple | try ret       |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | try ret       |  +  | tryCatch/elementary_non_address_state_mutability_try_returns.sol        |
|              | invalid  | catch param   |  +  | tryCatch/elementary_non_address_state_mutability_catch.sol              |
|              | multiple | file var      |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | file var      |  +  | parsing/elementary_non_address_state_mutability_file_var.sol            |
|              | multiple | state var     |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | state var     |     | parsing/elementary_non_address_state_mutability_state_variable.sol      |
|              | multiple | fn var        |  +  | parsing/multiple_statemutability_specifiers.sol                         |
|              | invalid  | fn var        |     | parsing/elementary_non_address_state_mutability_local.sol               |
|              | multiple | fn            |     | parsing/multiple_statemutability_specifiers.sol                         |
| view         | invalid  | mod param     |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | mod param     |  +  | modifiers/elementary_non_address_state_mutability_modifier_argument.sol |
|              | invalid  | fn par        |     | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | fn par        |  +  | parsing/elementary_non_address_state_mutability_argument.sol            |
|              | invalid  | fn ret        |     | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | fn ret        |  +  | parsing/elementary_non_address_state_mutability_return.sol              |
|              | invalid  | ev param      |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | ev param      |  +  | events/elementary_non_address_state_mutability_event_argument.sol       |
|              | invalid  | err param     |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | err param     |  +  | errors/elementary_non_address_state_mutability_error_argument.sol       |
|              | invalid  | try ret       |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | try ret       |  +  | tryCatch/elementary_non_address_state_mutability_try_returns.sol        |
|              | invalid  | catch param   |  +  | tryCatch/elementary_non_address_state_mutability_catch.sol              |
|              | invalid  | file var      |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | file var      |  +  | parsing/elementary_non_address_state_mutability_file_var.sol            |
|              | invalid  | state var     |     | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | state var     |  +  | parsing/elementary_non_address_state_mutability_state_variable.sol      |
|              | invalid  | fn var        |     | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | fn var        |  +  | parsing/elementary_non_address_state_mutability_local.sol               |
|              | multiple | fn            |     | parsing/multiple_statemutability_specifiers.sol                         |
| pure         | invalid  | mod param     |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | mod param     |  +  | modifiers/elementary_non_address_state_mutability_modifier_argument.sol |
|              | invalid  | fn par        |     | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | fn par        |  +  | parsing/elementary_non_address_state_mutability_argument.sol            |
|              | invalid  | fn ret        |     | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | fn ret        |  +  | parsing/elementary_non_address_state_mutability_return.sol              |
|              | invalid  | ev param      |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | ev param      |  +  | events/elementary_non_address_state_mutability_event_argument.sol       |
|              | invalid  | err param     |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | err param     |  +  | errors/elementary_non_address_state_mutability_error_argument.sol       |
|              | invalid  | try ret       |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | try ret       |  +  | tryCatch/elementary_non_address_state_mutability_try_returns.sol        |
|              | invalid  | catch param   |  +  | tryCatch/elementary_non_address_state_mutability_catch.sol              |
|              | invalid  | file var      |  +  | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | file var      |  +  | parsing/elementary_non_address_state_mutability_file_var.sol            |
|              | invalid  | state var     |     | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | state var     |  +  | parsing/elementary_non_address_state_mutability_state_variable.sol      |
|              | invalid  | fn var        |     | parsing/address_invalid_state_mutability.sol                            |
|              | invalid  | fn var        |  +  | parsing/elementary_non_address_state_mutability_local.sol               |
|              | multiple | fn            |     | parsing/multiple_statemutability_specifiers.sol                         |
| memory       | multiple | mod param     |  +  | modifiers/multiple_parameter_location.sol                               |
|              | multiple | fn par        |     | parsing/location_specifiers_for_params_multi.sol                        |
|              | multiple | fn ret        |  +  | parsing/location_specifiers_for_fn_returns_multi.sol                    |
|              | invalid  | ev param      |  +  | events/invalid_parameter_location.sol                                   |
|              | invalid  | err param     |  +  | errors/invalid_parameter_location.sol                                   |
|              | multiple | try ret       |  +  | tryCatch/multiple_returns_vars_and_catch_parameter_location.sol         |
|              | multiple | catch param   |  +  | tryCatch/multiple_returns_vars_and_catch_parameter_location.sol         |
|              | invalid  | file var      |  +  | parsing/location_specifiers_for_file_level_var.sol                      |
|              | invalid  | state var     |  +  | parsing/invalid_state_variable_location.sol                             |
|              | multiple | fn var        |     | parsing/location_specifiers_for_locals_multi.sol                        |
| calldata     | multiple | mod param     |  +  | modifiers/multiple_parameter_location.sol                               |
|              | multiple | fn par        |     | parsing/location_specifiers_for_params_multi.sol                        |
|              | multiple | fn ret        |  +  | parsing/location_specifiers_for_fn_returns_multi.sol                    |
|              | invalid  | ev param      |  +  | events/invalid_parameter_location.sol                                   |
|              | invalid  | err param     |  +  | errors/invalid_parameter_location.sol                                   |
|              | multiple | try ret       |  +  | tryCatch/multiple_returns_vars_and_catch_parameter_location.sol         |
|              | invalid  | catch param   |  +  | tryCatch/invalid_catch_parameter_location_calldata.sol                  |
|              | multiple | catch param   |  +  | tryCatch/multiple_returns_vars_and_catch_parameter_location.sol         |
|              | invalid  | file var      |  +  | parsing/location_specifiers_for_file_level_var.sol                      |
|              | invalid  | state var     |  +  | parsing/invalid_state_variable_location.sol                             |
|              | multiple | fn var        |     | parsing/location_specifiers_for_locals_multi.sol                        |
| storage      | multiple | mod param     |  +  | modifiers/multiple_parameter_location.sol                               |
|              | multiple | fn par        |     | parsing/location_specifiers_for_params_multi.sol                        |
|              | multiple | fn ret        |  +  | parsing/location_specifiers_for_fn_returns_multi.sol                    |
|              | invalid  | ev param      |  +  | events/invalid_parameter_location.sol                                   |
|              | invalid  | err param     |  +  | errors/invalid_parameter_location.sol                                   |
|              | invalid  | try ret       |  +  | tryCatch/invalid_returns_var_location.sol                               |
|              | multiple | try ret       |  +  | tryCatch/multiple_returns_vars_and_catch_parameter_location.sol         |
|              | invalid  | catch param   |  +  | tryCatch/invalid_catch_parameter_location_storage.sol                   |
|              | multiple | catch param   |  +  | tryCatch/multiple_returns_vars_and_catch_parameter_location.sol         |
|              | invalid  | file var      |  +  | parsing/location_specifiers_for_file_level_var.sol                      |
|              | multiple | state var     |     | parsing/location_specifiers_for_state_variables_multi.sol               |
|              | multiple | fn var        |     | parsing/location_specifiers_for_locals_multi.sol                        |
| experimental | multiple |               |  +  | pragma/experimental_multiple_experimental.sol                           |

**Comment by vlad-khramov** [comment]
rebased bcs of docs build error (since there were fixed in develop) but did not help

**Comment by nikola-matic** [comment]
@quantum13 feel free to rebase again, the docs (and other) build failures should be fixed now.

**Comment by quantum13** [comment]
@cameel @nikola-matic could you please check this PR

**Review by cameel — DISMISSED** [review]
Wow, that's excellent job. Thank you! I rarely see contributions that are this thorough and don't have even one small issue. Honestly, I expected just a few tests to check the most obvious cases for each keyword but you went above and beyond that. And looks like it was perfectly justified - we did have another bug related to `indexed`. Kudos for finding and fixing it.

I rebased the PR and added some cosmetic tweaks like whitespace corrections and typo fixes in a couple of test names so that we can merge it as soon as tests pass.

**Review by cameel — APPROVED** [review]

**Comment by vlad-khramov** [comment]
thank you :)

**Pull request merged**